src/core/thread/network_data_leader_ftd.cpp

```diff
@@ -855,7 +854,8 @@ Error Leader::AddBorderRouter(const BorderRouterTlv &aBorderRouter, PrefixTlv &a
         // Allocate a Context ID first. This ensure that if we cannot
         // allocate, we fail and exit before potentially inserting a
         // Border Router sub-TLV.
-        SuccessOrExit(error = AllocateContextId(contextId));
+        contextId = mContextIds.AllocateNewId();
```

Comment:
Should we free the context id if any error occurs below?

Reply:
Good point! 
I changed the logic here. The method is now called `GetUnallocatedId()` and it just find an unallocated ID but won't change its state. We call `MarkAsInUse()` below after we insert the `ContextTlv` in netdata.